src/main/kotlin/com/milaboratory/mixcr/cli/CommandBuildLibrary.kt

```diff
@@ -0,0 +1,394 @@
+            Diversity -> dGenes
+            Joining -> jGenes
+            Constant -> cGenes
+        }
+        if (input == null) {
+            null
+        } else if (input.species != null) {
+            fromKnownSpecies(gt, input.species!!)
+        } else {
+            val fasta = input.fasta!!
+            fromFasta(gt, fasta.geneFeatureOps!!, fasta.path)
+        }
+    }
+
+    private fun tmpLibraryName(geneType: GeneType) =
+        output!!.absolutePathString().replace(".json", ".${geneType.letter}.json")
+
+    private fun fromKnownSpecies(geneType: GeneType, species: String) = run {
+        val chains = Chains.parse(chain)
+        val data = VDJCLibraryRegistry.getDefaultLibrary(species).data
+        val dataFiltered =
+            data.copy(
+                genes = data.genes.filter { it.geneType == geneType && it.chains.intersects(chains) },
+                speciesNames = listOf(this.species),
+                taxonId = java.lang.Long.parseLong(taxonId)
+            )
+        val libName = tmpLibraryName(geneType)
+        writeToFile(listOf(dataFiltered), libName, false)
+        libName
+    }
+
+    private fun logCmd(cmd: List<String>) {
+        logger.debug { "Running:\nrepseqio ${cmd.joinToString(" ")}" }
+    }
+
+    private fun fromFasta(
+        geneType: GeneType,
+        geneFeature: String,
+        fasta: Path
+    ) = run {
+        val libName = tmpLibraryName(geneType)
+
+        val ffCmd = """
+            fromFasta
+            --taxon-id $taxonId
+            --species-name $species
+            --gene-type ${geneType.letter}
+            --chain $chain
+            --gene-feature $geneFeature
+            --name-index 0
+            -f
+            ${fasta.absolutePathString()}
+            $libName
+        """.trimIndent()
+            .replace("\n", " ")
+            .split(" ")
+            .filter { it.isNotBlank() }
+        logCmd(ffCmd)
+        io.repseq.cli.Main.main(ffCmd.toTypedArray())
+
+        if (!doNotInferPoints) {
+            val ipCmd = listOf("inferPoints", "-g", geneFeature, "-f", libName, libName)
+            logCmd(ipCmd)
+            io.repseq.cli.Main.main(ipCmd.toTypedArray())
+        }
+
+        val coCmd = listOf("compile", "--do-not-compress", "-f", libName, libName)
+        logCmd(coCmd)
+        io.repseq.cli.Main.main(coCmd.toTypedArray())
+
+        libName
+    }
+
+    override fun run1() {
+        val libName = output!!.absolutePathString()
+        val meCmd = mutableListOf("merge", "-f")
+        meCmd += GeneType.values().mapNotNull { mkLibraryForGeneType(it) }
+        meCmd += libName
+        if (output?.endsWith("gz") == true)
+            meCmd += "--compress"
+        logCmd(meCmd)
+        io.repseq.cli.Main.main(meCmd.toTypedArray())
+
+        if (logger.verbose) {
+            val deCmd = listOf("debug", libName)
+            io.repseq.cli.Main.main(deCmd.toTypedArray())
+        }
+
+        if (!keepIntermediateFiles)
```

Comment:
they will not be deleted in case of any exception. Use try finaly

Reply:
👍

Reply:
done